server/app/auth/CiviFormProfile.java

```diff
@@ -44,6 +45,27 @@ public CiviFormProfile(
 
   /** Get the latest {@link Applicant} associated with the profile. */
   public CompletableFuture<Applicant> getApplicant() {
+    if (profileData.containsAttribute(ProfileFactory.APPLICANT_ID_ATTRIBUTE_NAME)) {
+      return supplyAsync(
+          () -> {
+            long applicantId =
+                profileData.getAttribute(ProfileFactory.APPLICANT_ID_ATTRIBUTE_NAME, Long.class);
+            Applicant applicant = new Applicant();
+            applicant.id = applicantId;
+            try {
+              applicant.refresh();
+            } catch (EntityNotFoundException e) {
+              // Unlike AccountNotFoundException, ApplicantNotFoundException is not a
+              // RuntimeException.
+              throw new RuntimeException(new ApplicantNotFoundException(applicantId));
+            }
+            return applicant;
```

Comment:
suggestion (blocking): use AccountRepository.lookupApplicant instead

In general we're trying to move toward encapsulating all query logic into repository classes where we'll have an easier time optimizing it and controlling how data is loaded

Reply:
Sigh, yes, I should remember this by now, instead of just borrowing a pattern from nearby. Done